Hmm it seems like it gets through a few hundred samples and then vllm crashes with this error:

<details>
<summary> AssertionError: aligned_length is not compatible with eagle now </summary>

```
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112] EngineCore encountered a fatal error.
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112] Traceback (most recent call last):
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]   File "/home/fynnsu/repos/vllm-project/vllm/vllm/v1/engine/core.py", line 1103, in run_engine_core
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]     engine_core.run_busy_loop()
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]   File "/home/fynnsu/repos/vllm-project/vllm/vllm/v1/engine/core.py", line 1144, in run_busy_loop
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]     self._process_engine_step()
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]   File "/home/fynnsu/repos/vllm-project/vllm/vllm/v1/engine/core.py", line 1183, in _process_engine_step
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]     outputs, model_executed = self.step_fn()
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]   File "/home/fynnsu/repos/vllm-project/vllm/vllm/v1/engine/core.py", line 451, in step_with_batch_queue
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]     scheduler_output = self.scheduler.schedule()
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]   File "/home/fynnsu/repos/vllm-project/vllm/vllm/v1/core/sched/scheduler.py", line 613, in schedule
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]     self.kv_cache_manager.get_computed_blocks(request)
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]   File "/home/fynnsu/repos/vllm-project/vllm/vllm/v1/core/kv_cache_manager.py", line 203, in get_computed_blocks
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]     self.coordinator.find_longest_cache_hit(
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]   File "/home/fynnsu/repos/vllm-project/vllm/vllm/v1/core/kv_cache_coordinator.py", line 514, in find_longest_cache_hit
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]     hit_blocks = manager_cls.find_longest_cache_hit(
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]   File "/home/fynnsu/repos/vllm-project/vllm/vllm/v1/core/single_type_kv_cache_manager.py", line 568, in find_longest_cache_hit
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112]     assert kv_cache_spec.block_size == alignment_tokens, (
(EngineCore pid=2643696) ERROR 04-10 00:22:31 [core.py:1112] AssertionError: aligned_length is not compatible with eagle now
(APIServer pid=2643277) ERROR 04-10 00:22:31 [async_llm.py:701] AsyncLLM output_handler failed.
(APIServer pid=2643277) ERROR 04-10 00:22:31 [async_llm.py:701] Traceback (most recent call last):
(APIServer pid=2643277) ERROR 04-10 00:22:31 [async_llm.py:701]   File "/home/fynnsu/repos/vllm-project/vllm/vllm/v1/engine/async_llm.py", line 657, in output_handler
(APIServer pid=2643277) ERROR 04-10 00:22:31 [async_llm.py:701]     outputs = await engine_core.get_output_async()
(APIServer pid=2643277) ERROR 04-10 00:22:31 [async_llm.py:701]   File "/home/fynnsu/repos/vllm-project/vllm/vllm/v1/engine/core_client.py", line 1000, in get_output_async
(APIServer pid=2643277) ERROR 04-10 00:22:31 [async_llm.py:701]     raise self._format_exception(outputs) from None
(APIServer pid=2643277) ERROR 04-10 00:22:31 [async_llm.py:701] vllm.v1.engine.exceptions.EngineDeadError: EngineCore encountered an issue. See stack trace (above) for the root cause.
(Worker_TP1 pid=2644262) INFO 04-10 00:22:31 [multiproc_executor.py:881] WorkerProc shutting down.
(Worker_TP0 pid=2644261) INFO 04-10 00:22:31 [multiproc_executor.py:784] Parent process exited, terminating worker queues
(APIServer pid=2643277) INFO:     127.0.0.1:58436 - "POST /v1/chat/completions HTTP/1.1" 500 Internal Server Error
(Worker_TP0 pid=2644261) INFO 04-10 00:22:31 [multiproc_executor.py:881] WorkerProc shutting down.
(APIServer pid=2643277) INFO:     Shutting down
(APIServer pid=2643277) INFO:     Waiting for application shutdown.
(APIServer pid=2643277) INFO:     Application shutdown complete.
(APIServer pid=2643277) INFO:     Finished server process [2643277]
```
</details>

The failing assert is under a `use_eagle` branch of the `SlidingWindowManager.find_longest_cache_hit` method.

I tried running gsm8k eval on `RedHatAI/gpt-oss-20b-speculator.eagle3` in case there is an issue with sliding windows + eagle on latest main, but that passed.

--- a/vllm/config/speculative.py
+++ b/vllm/config/speculative.py
@@ -825,6 +825,7 @@ class SpeculativeConfig:
             "kimi_k2",
             "kimi_k25",
             "minimax_m2",
+            "gemma4",
         ]
         if (
             self.method in ("eagle3", "extract_hidden_states", "dflash")

--- a/vllm/model_executor/models/gemma4.py
+++ b/vllm/model_executor/models/gemma4.py
@@ -60,7 +60,13 @@ from vllm.model_executor.model_loader.weight_utils import (
 from vllm.sequence import IntermediateTensors
 from vllm.v1.attention.backends.utils import KVSharingFastPrefillMetadata
 
-from .interfaces import MixtureOfExperts, SupportsLoRA, SupportsPP
+from .interfaces import (
+    EagleModelMixin,
+    MixtureOfExperts,
+    SupportsEagle3,
+    SupportsLoRA,
+    SupportsPP,
+)
 from .utils import (
     AutoWeightsLoader,
     extract_layer_index,
@@ -838,7 +844,7 @@ class Gemma4CrossDecoderLayers(nn.Module):
 @support_torch_compile(
     enable_if=lambda vllm_config: not vllm_config.cache_config.kv_sharing_fast_prefill
 )
-class Gemma4Model(nn.Module):
+class Gemma4Model(nn.Module, EagleModelMixin):
     def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
         super().__init__()
         config = _get_text_config(vllm_config.model_config.hf_config)
@@ -1168,7 +1174,7 @@ class Gemma4Model(nn.Module):
         inputs_embeds: torch.Tensor | None = None,
         per_layer_inputs: torch.Tensor | None = None,
         **kwargs,
-    ) -> torch.Tensor | IntermediateTensors:
+    ) -> torch.Tensor | IntermediateTensors | tuple[torch.Tensor, list[torch.Tensor]]:
         if self.fast_prefill_enabled:
             hidden_states = self.fast_prefill_forward(
                 input_ids,
@@ -1204,6 +1210,7 @@ class Gemma4Model(nn.Module):
             residual = intermediate_tensors["residual"]
             per_layer_inputs = intermediate_tensors.get("per_layer_inputs")
 
+        aux_hidden_states = self._maybe_add_hidden_state([], 0, hidden_states, residual)
         for layer_idx, layer in enumerate(
             islice(self.layers, self.start_layer, self.end_layer)
         ):
@@ -1222,6 +1229,9 @@ class Gemma4Model(nn.Module):
                 per_layer_input=layer_per_input,
                 **kwargs,
             )
+            self._maybe_add_hidden_state(
+                aux_hidden_states, layer_idx + 1, hidden_states, residual
+            )
         if not get_pp_group().is_last_rank:
             return IntermediateTensors(
                 {
@@ -1236,6 +1246,9 @@ class Gemma4Model(nn.Module):
             hidden_states = self.norm(hidden_states)
         else:
             hidden_states, _ = self.norm(hidden_states, residual)
+
+        if len(aux_hidden_states) > 0:
+            return hidden_states, aux_hidden_states
         return hidden_states
 
     def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
@@ -1381,7 +1394,9 @@ class Gemma4Model(nn.Module):
         return loaded_params
 
 
-class Gemma4ForCausalLM(nn.Module, SupportsLoRA, SupportsPP, MixtureOfExperts):
+class Gemma4ForCausalLM(
+    nn.Module, SupportsLoRA, SupportsPP, MixtureOfExperts, SupportsEagle3
+):
     # Note: qkv_proj packing applies to non-k_eq_v layers (sliding
     # attention and full attention without k_eq_v). k_eq_v layers use
     # separate q_proj + k_proj without packing.
@@ -1463,7 +1478,7 @@ class Gemma4ForCausalLM(nn.Module, SupportsLoRA, SupportsPP, MixtureOfExperts):
         intermediate_tensors: IntermediateTensors | None = None,
         inputs_embeds: torch.Tensor | None = None,
         **kwargs,
-    ) -> torch.Tensor | IntermediateTensors:
+    ) -> torch.Tensor | IntermediateTensors | tuple[torch.Tensor, list[torch.Tensor]]:
         hidden_states = self.model(
             input_ids, positions, intermediate_tensors, inputs_embeds, **kwargs
         )

--- a/vllm/model_executor/models/gemma4_mm.py
+++ b/vllm/model_executor/models/gemma4_mm.py
@@ -64,7 +64,12 @@ from vllm.multimodal.processing.processor import (
 from vllm.sequence import IntermediateTensors
 from vllm.utils.tensor_schema import TensorSchema, TensorShape
 
-from .interfaces import MultiModalEmbeddings, SupportsMultiModal, SupportsPP
+from .interfaces import (
+    MultiModalEmbeddings,
+    SupportsEagle3,
+    SupportsMultiModal,
+    SupportsPP,
+)
 from .utils import (
     AutoWeightsLoader,
     WeightsMapper,
@@ -845,7 +850,12 @@ class Gemma4MultimodalEmbedder(nn.Module):
     info=Gemma4ProcessingInfo,
     dummy_inputs=Gemma4DummyInputsBuilder,
 )
-class Gemma4ForConditionalGeneration(nn.Module, SupportsMultiModal, SupportsPP):
+class Gemma4ForConditionalGeneration(
+    nn.Module,
+    SupportsMultiModal,
+    SupportsPP,
+    SupportsEagle3,
+):
     packed_modules_mapping = {
         "qkv_proj": [
             "q_proj",

